File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-10.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while(l < r){
            int mid = l + (r - l) / 2;
            if(nums[r] < nums[mid]){
                l = mid + 1;
            }
            else{
                r = mid;
            }
        }

        int pivot = l;
        l = 0, r = nums.size() - 1;
        if(pivot == 0){}
        else if(target >= nums[0]) r = pivot - 1;
        else l = pivot;

        while(l <= r){
            int mid = l + (r - l) / 2;
            if(target == nums[mid]) return mid;
            else if(target > nums[mid]) l = mid + 1;
            else r = mid - 1;
        }

        return -1;
    }
};

File: Data Structures & Algorithms/three-integer-sum/submission-10.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # nums[j] + nums[k] == -1 * nums[i]
        nums.sort()
        res = []
        for i in range(len(nums)):
            if nums[i] > 0:
                break
            
            if i > 0 and nums[i] == nums[i-1]:
                continue
            
            j = i + 1
            k = len(nums) - 1
            target = nums[i] * -1
            while j < k:
                sum = nums[j] + nums[k]
                if sum == target:
                    res.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j-1]:
                        j += 1
                elif sum < target:
                    j += 1
                else:
                    k -= 1
        return res
